### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by tiserett**
[DEMO LINK](https://tiserett.github.io/react_todo-app-with-api/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,223 +1,220 @@
...
+import { FilterType } from './types/FilterStatus';
+import { Todo } from './types/Todo';
+import { Error } from './types/Error';
 
 export const App: React.FC = () => {
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const user = useContext(AuthContext);
   const newTodoField = useRef<HTMLInputElement>(null);
+  const [error, setError] = useState<Error | null>(null);
+  const [filterType, setFilterType] = useState('all');
```
Comment (sergii-nosachenko) [inline src/App.tsx:28]:
```suggestion
  const [filterType, setFilterType] = useState(FilterState.All);
```

Not every string can be a filter value

File: src/components/Footer/Footer.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import { Props } from './Footer.props';
+
+export const Footer: React.FC<Props> = ({
+  filterType,
+  handleFilterType,
+  todos,
+  deleteCompleted,
+}) => {
+  const todosLeft = todos.filter(todo => !todo.completed).length;
+  const todosCompleted = todos.filter(todo => todo.completed).length;
```
Comment (sergii-nosachenko) [inline src/components/Footer/Footer.tsx:13]:
```suggestion
  const todosCompleted = todos.length - todosLeft;
```

Redundant filter operation (simplify complexity)

File: src/components/TodoItem/TodoItem.props.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { Todo } from '../../types/Todo';
+
+export type Props = {
+  todo: Todo;
+  onDelete: (todoId: number) => void;
+  selectedTodos: number[];
+  setSelectedTodos: (todoId: number[]) => void;
+  onUpdate: (todoId: number, data: Partial<Todo>) => void;
```
Comment (sergii-nosachenko) [inline src/components/TodoItem/TodoItem.props.ts:8]:
```suggestion
  onUpdate: (todoId: number, data: RequireAtLeastOne<Todo, 'completed' | 'title'>) => void;
```

Are you sure that it's safe to make all Todo properties optional? Now you can pass even empty Todo. Better use another approach.

File: src/components/TodoItem/TodoItem.tsx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        : (
+          <>
+            <span
+              data-cy="TodoTitle"
+              className="todo__title"
+              onDoubleClick={() => {
+                setIsClicked(true);
+                setSelectedTodo(todo.id);
+                setNewTodoTitle(todo.title);
+              }}
```
Comment (sergii-nosachenko) [inline src/components/TodoItem/TodoItem.tsx:112]: Create hadler for this complex operation

**Review by sergii-nosachenko — CHANGES_REQUESTED** [review]
Very good work, Max! 👍🐱‍👤
But I've found a bunch of mistakes need to be fixed. Feel free to discuss any issue with me.

**Review by sergii-nosachenko — APPROVED** [review]
Great job 🔥

File: src/App.tsx [hunk 5]
```diff
@@ -1,223 +1,232 @@
...
+  const [selectedTodos, setSelectedTodos] = useState<number[]>([]);
+  const [toggle, setToggle] = useState(true);
+  const [selectedTodo, setSelectedTodo] = useState(0);
+
+  if (error) {
+    setTimeout(() => {
+      setError(null);
+      setIsAdding(false);
+    }, 3000);
+  }
```
Comment (AlinaMatuschak) [inline src/App.tsx]: Maybe will be better to create setTimeout on the ErrorNotification and clear timeout on unmount (use useEffect)

File: src/App.tsx [hunk 6]
```diff
@@ -1,223 +1,232 @@
...
+
+  let userId = 0;
+
+  if (user?.id) {
+    userId = user?.id;
+  }
```
Comment (AlinaMatuschak) [inline src/App.tsx]:
```suggestion
  const userId = user?.id || 0;
```

File: src/App.tsx [hunk 7]
```diff
@@ -1,223 +1,232 @@
...
+
+  const filteredTodos = todos.filter(todo => {
```
Comment (AlinaMatuschak) [inline src/App.tsx]: useMemo

File: src/App.tsx [hunk 8]
```diff
@@ -1,223 +1,232 @@
...
+        return todo;
+
+      case FilterType.Active:
+        return !todo.completed;
+
+      case FilterType.Completed:
+        return todo.completed;
+
+      default:
+        return null;
```
Comment (AlinaMatuschak) [inline src/App.tsx]:
```suggestion
    switch (filterType) {
      case FilterType.Active:
        return !todo.completed;

      case FilterType.Completed:
        return todo.completed;

      default:
        return true;
```

File: src/App.tsx [hunk 9]
```diff
@@ -1,223 +1,232 @@
...
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
-    }
+    inputField.current?.focus();
+
+    getTodos(userId)
+      .then(setTodos)
+      .catch(() => setError(Error.LOADING));
```
Comment (AlinaMatuschak) [inline src/App.tsx]: Let’s use async await. It gives you a simpler way to work with asynchronous promise-based code

File: src/App.tsx [hunk 10]
```diff
@@ -1,223 +1,232 @@
...
+      return;
+    }
+
+    setIsAdding(true);
+    setSelectedTodos([userId]);
+
+    try {
+      const newTodo = await createTodo(userId, title);
+
+      setTodos([...todos, newTodo]);
```
Comment (AlinaMatuschak) [inline src/App.tsx:91]: If the new state is computed using the previous state, you can pass a function to setTodos

File: src/App.tsx [hunk 11]
```diff
@@ -1,223 +1,232 @@
...
+    getTodos(userId)
+      .then(setTodos)
+      .catch(() => setError(Error.LOADING));
   }, []);
 
+  useEffect(() => {
+    inputField.current?.focus();
+  }, [todos]);
+
+  const handleSubmit = async (event: FormEvent) => {
```
Comment (AlinaMatuschak) [inline src/App.tsx]: useCallback

File: src/App.tsx [hunk 12]
```diff
@@ -1,223 +1,232 @@
...
+    try {
+      await removeTodo(todoId);
+
+      setTodos(todos.filter(todo => todo.id !== todoId));
+    } catch {
+      setError(Error.DELETING);
+    }
+  };
+
+  const completedTodos = todos.filter(todo => todo.completed);
```
Comment (AlinaMatuschak) [inline src/App.tsx]: useMemo

File: src/App.tsx [hunk 13]
```diff
@@ -1,223 +1,232 @@
...
+      setError(Error.UPDATING);
+    }
+
+    setSelectedTodos([]);
+  };
+
+  const handleToggle = async () => {
+    setSelectedTodos(toggle
+      ? [...todos].filter(todo => !todo.completed).map(todo => todo.id)
+      : [...completedTodos].map(todo => todo.id));
```
Comment (AlinaMatuschak) [inline src/App.tsx:151]: use functions on setSelectedTodos

Comment (AlinaMatuschak) [inline src/App.tsx:151]: do you need to create here copy completedTodos? Map returns new array

File: src/components/ErrorNotification/ErrorNotification.tsx [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+}) => {
+  return (
+    <div
+      data-cy="ErrorNotification"
+      className={classnames(
+        'notification is-danger is-light has-text-weight-normal',
+        { hidden: !errorText },
+      )}
+    >
+      <>
```
Comment (AlinaMatuschak) [inline src/components/ErrorNotification/ErrorNotification.tsx]: you can delete fragment

File: src/components/Footer/Footer.tsx [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+import { Props } from './Footer.props';
+
+export const Footer: React.FC<Props> = ({
+  filterType,
+  handleFilterType,
+  todos,
+  deleteCompleted,
+}) => {
+  const todosLeft = todos.filter(todo => !todo.completed).length;
```
Comment (AlinaMatuschak) [inline src/components/Footer/Footer.tsx]: useMemo

File: src/components/Footer/Footer.tsx [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+      </nav>
+
+      <button
+        data-cy="ClearCompletedButton"
+        type="button"
+        className="todoapp__clear-completed"
+        onClick={deleteCompleted}
+      >
+        {todosCompleted > 0 && 'Clear completed'}
+      </button>
```
Comment (AlinaMatuschak) [inline src/components/Footer/Footer.tsx:68]: the same as in prev work

File: src/components/TodoItem/TodoItem.tsx [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+            <input
+              data-cy="TodoTitleField"
+              type="text"
+              ref={newTodoField}
+              className="todo__title-field"
+              value={newTodoTitle}
+              placeholder="If your todo is empty, it will be deleted"
+              onChange={event => handleTitleChange(event)}
+              onBlur={handleBlur}
+              onKeyDown={event => handleKeyPress(event)}
```
Comment (AlinaMatuschak) [inline src/components/TodoItem/TodoItem.tsx:119]:
```suggestion
              onChange={handleTitleChange}
              onBlur={handleBlur}
              onKeyDown={handleKeyPress}
```

**Review by AlinaMatuschak — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 18]
```diff
@@ -1,223 +1,237 @@
...
+    } catch {
+      setError(Error.ADDING);
+    }
+
+    setTitle('');
+    setIsAdding(false);
+    setSelectedTodos([]);
+  }, [title]);
+
+  const deleteTodo = async (todoId: number) => {
```
Comment (AlinaMatuschak) [inline src/App.tsx:101]: useCallback

File: src/App.tsx [hunk 19]
```diff
@@ -1,223 +1,237 @@
...
+    setSelectedTodos([todoId]);
+
+    try {
+      await removeTodo(todoId);
+
+      setTodos(todos.filter(todo => todo.id !== todoId));
```
Comment (AlinaMatuschak) [inline src/App.tsx:107]: If the new state is computed using the previous state, you can pass a function to setTodos

File: src/App.tsx [hunk 20]
```diff
@@ -1,223 +1,237 @@
...
+    return todos.filter(todo => todo.completed);
+  }, [todos]);
+
+  const deleteCompletedTodos = useCallback(async () => {
+    setSelectedTodos([...completedTodos].map(todo => todo.id));
+
+    try {
+      await Promise.all(completedTodos.map(todo => removeTodo(todo.id)));
+
+      setTodos([...todos.filter(todo => !todo.completed)]);
```
Comment (AlinaMatuschak) [inline src/App.tsx:123]: and here

File: src/App.tsx [hunk 21]
```diff
@@ -1,223 +1,237 @@
...
+    } catch {
+      setError(Error.DELETING);
+      setSelectedTodos([]);
+    }
+  }, [completedTodos]);
+
+  const handleTodoUpdate = async (todoId: number, data: Partial<Todo>) => {
```
Comment (AlinaMatuschak) [inline src/App.tsx:130]: useCallback

File: src/App.tsx [hunk 22]
```diff
@@ -1,223 +1,237 @@
...
+    setSelectedTodos([todoId]);
+
+    try {
+      const newTodo = await updateTodo(todoId, data);
+
+      setTodos(todos.map(todo => (
```
Comment (AlinaMatuschak) [inline src/App.tsx:136]: use function in setTodos

File: src/App.tsx [hunk 23]
```diff
@@ -1,223 +1,237 @@
...
+          : todo
+      )));
+    } catch {
+      setError(Error.UPDATING);
+    }
+
+    setSelectedTodos([]);
+  };
+
+  const handleToggle = async () => {
```
Comment (AlinaMatuschak) [inline src/App.tsx:148]: use Callback

File: src/api/todos.ts [hunk 24]
```diff
@@ -5,4 +5,18 @@ export const getTodos = (userId: number) => {
...
+  return client.post<Todo>('/todos', {
+    userId,
+    completed: false,
+    title,
+  });
+};
+
+export const removeTodo = (todoID: number) => {
+  return client.delete(`/todos/${todoID}`);
+};
```
Comment (AlinaMatuschak) [inline src/api/todos.ts:18]: add type for the result

File: src/components/TodoItem/TodoItem.tsx [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+  onUpdate,
+  selectedTodo,
+  setSelectedTodo,
+  todos,
+}) => {
+  const newTodoField = useRef<HTMLInputElement>(null);
+  const [isClicked, setIsClicked] = useState(false);
+  const [newTodoTitle, setNewTodoTitle] = useState('');
+
+  const handleTitleUpdate = () => {
```
Comment (AlinaMatuschak) [inline src/components/TodoItem/TodoItem.tsx:26]: useCallback 😭

**Review by AlinaMatuschak — APPROVED** [review]
Nice 👍 . But fix comments 🙏

make sure that it looks like on mate demo:
Screenshot 2022-10-05 at 20 32 41

mate demo:
Screenshot 2022-10-05 at 20 33 31

File: src/components/TodoItem/TodoItem.tsx (hunk 25 above)
Comment (tiserett) [reply src/components/TodoItem/TodoItem.tsx:26]: is it really useful to use hook useCallback without optimized child components? I watched few videos about it, and now I'm not sure about covering every function with it

Comment (AlinaMatuschak) [reply src/components/TodoItem/TodoItem.tsx:26]: yeap. If you don’t use useCallback, your function will recreate each time when something in the component update

**Pull request closed without merging**